src/builder.test.ts

```diff
@@ -82,6 +82,47 @@ describe("ArrayResult.grab/UnknownResult.grab/EntityResult.grab", () => {
     expect(data).toEqual({ name: null });
   });
 
+  it("can handle conditional selections without base query", async () => {
+    const { data, query } = await runPokemonQuery(
+      q("*")
+        .filter("_type == 'pokemon'")
+        .slice(0, 3)
+        .grab(
+          {},
+          {
+            "name == 'Charmander'": {
+              name: q.literal("Charmander"),
+              hp: ["base.HP", q.number()],
+            },
+            "name == 'Bulbasaur'": {
+              name: q.literal("Bulbasaur"),
+              attack: ["base.Attack", q.number()],
+            },
+          }
+        )
+    );
+
+    expect(query).toBe(
+      `*[_type == 'pokemon'][0..3]{...select(name == 'Charmander' => { name, "hp": base.HP }, name == 'Bulbasaur' => { name, "attack": base.Attack })}`
+    );
+
+    invariant(data);
+    expect(data[0]).toEqual({ name: "Bulbasaur", attack: 49 });
+    expect(data[1]).toEqual({});
```

Comment:
I think this is the thing that is getting me hung up. If there's an empty base selection, and our type looks something like `({ name: "Charmander", ... } | { name: "Bulbasaur", ... })[]`, then having an empty object is a "wrong" according to our type.

But in groq, I believe you can have an empty projection and it'll just return an empty object. So for example, this query:

```
*[_type == 'pokemon'][0..3]{...select(name == 'Charmander' => { name, "hp": base.HP }, name == 'Bulbasaur' => { name, "attack": base.Attack })}
```

does indeed pass the tests you have here, _including_ the empty-object one – since that projection will just select nothing on everything but Charmander and Bulbasaur.

Still trying to wrap my head around this, but I think limiting our types on this might end up causing a mismatch with how groq behaves – which would be confusing.

Any thoughts on this? It's not yet clear to me how to proceed on this.

Reply:
Jep, that's true. Also don't know, how to proceed either. Then, I think, the PR can be closed, and it's up to the user, to check, whether the empty object is matching, or not.

Though, on the other side, if there is an empty object in the response, I don't think there is a big case for any user to show an empty tile/card/list-item/whatever, so my opinion on this would be to just add a third argument to the conditionals, and just set a default option.

I am thinking of an option named `includeEmptyObject`, which is then conditionally giving the user at least the type. But then, we would need to filter the empty object on our side.